### DISCUSSION TITLE
YAML input for Sage Attention node?

### DISCUSSION THREAD

**Discussion by EnragedAntelope** [blepping/ComfyUI-bleh#12] [Q&A] [answered]
Apologies if I missed it but I read the readme and didn't see info on what the YAML input is expecting specifically for the Sage Attention new nodes? Please share guidance on how best to use this... I tried SageAttn2 with CogVideoX via Kijai's nodes there and am very impressed at the speedup so looking forward to using it more. Thanks!

![image](https://github.com/user-attachments/assets/a614714e-93ec-491a-b001-3933ee4ac2dc)

**Comment by blepping** [comment, marked as answer]
No need to apologize. There isn't really specific documentation for using it as it is an advanced feature that allows you to directly pass (or override) arguments for the SageAttention function(s) which are provided by the SageAttention package. So basically, if something in the package changes, that would also change what you can/should pass.

This reply got pretty long so if you want a **TL;DR**: If you don't already know you need to change something there then it's probably not something you need to worry about/mess with. If you still want to know what you _could_ use it to do, then read on!

You pretty much need to look at the source code for SageAttention to see what it parameters you can change: https://github.com/thu-ml/SageAttention/blob/main/sageattention/core.py

The `yaml_parameters` input can be connected to a multiline text widget or converted into one in the node itself. Here is an example of parameters you could possibly pass (to the v2.0.0 beta version):

![image](https://github.com/user-attachments/assets/cf28df85-980e-4b92-80df-66967afafd90)

Most keys you specify will just be passed to the function directly but there are a couple exceptions. `sageattn_function` can be used to select a function that the SageAttention package provides by name so you could use it to override the kernel SageAttention automatically selected for your architecture. _Should_ you do this? Likely not, but you can if you want to. `sm_scale` is the softmax scale (don't ask me exactly what it does because I don't know!) which normally automatically gets set to the `tensor_head_dimension ** -0.5` but you can mess with it to get a different effect.

Just for example, the only head dimension in SDXL that SageAttention currently supports is 64, `64 ** -0.5 == 0.125` so for SDXL values close to `0.125` will be roughly what would be auto-selected. Raising/lowering it a bit seems to produce effects a bit like CFG, a bit like ComfyUI's experimental attention multiplier nodes. YAML parameters currently has one other special key you can pass, and that is setting `sm_scale_<head dim here>`, for example `sm_scale_64: 0.125` would use that scale specifically for attentions with head dimension 64 and you could set `sm_scale_96: <other value>` if you wanted to use something else for a different head dimension.

Note that I'm not saying it's better to change that or anything, just that you _can_ for a different effect (I like having knobs to turn!). Passing incorrect parameters, forcing a kernel your architecture doesn't expect, etc are all things that are likely to cause SageAttention to have unintended behavior/perform worse/maybe even crash ComfyUI or Torch. In other words, if you mess with that then you shouldn't bug the SageAttention authors if you run into problems.

Hope this helps answer your question!

**Comment by blepping** [comment]
By the way, I recommend using `SageAttentionSampler` rather than `GlobalSageAttention` when possible. You're less likely to run into problems (since my implementation patches/replaces ComyfUI's default optimized attention implementation). Also, the global node isn't _really_ a model patch, so for example if you were to bypass it that wouldn't actually disable SageAttention: you have to actually make sure the node runs with your changed settings (including the enabled/disabled toggle) for them to have effect.

The sampler specific version has less sharp edges.

**Comment by EnragedAntelope** [comment]
Thank you for the helpful explanation. 
I just had a chance to retry this. I used the sampler this time and it works fine and did not force me to add YAML parameters, so I skipped that entirely. However I saw no noticeable speed difference with flux at 896x1152, is that expected? It sounds like it would only kick in with a hiresfix sort of node I suppose?

**Reply by blepping** [reply]
>  I used the sampler this time and it works fine and did not force me to add YAML parameters, so I skipped that entirely. 

Oh, sorry and thanks for point that out. `yaml_parameters` was supposed to be optional for both nodes. Should be fixed now.

> However I saw no noticeable speed difference with flux at 896x1152, is that expected? 

As far as I know, SDXL is the only image model with compatible attentions (some video models _might_ be, I haven't tested them). It didn't seem to do anything for Flux when I tested, it definitely doesn't work with SD15.

You can try that `sageattn_verbose: true` parameter show in the screenshot above if you want to double check it. It should spam your console with messages like:

```plaintext
>>> SAGE(False): reshape=..... blah blah blah, dim_head=<some number>
```

`SAGE(False)` means the attention was not supported so it just got delegated to non-SageAttention. This is probably what you'll see, if you see `SAGE(True)` that means it's actually using SageAttention. Looking at `dim_head` might also be useful: SageAttention only supports 64, 96 and 128 at the moment. They say they're working on supporting other dimensions, so once that is added I will update the node to be able to use them. Unfortunately, it's not really something I can do myself: have to wait for them to support it, if they ever do.

**Comment by blepping** [comment]
@EnragedAntelope I just merged some changes for the bleh SageAttention support that should make it work for Flux now (in my limited testing I didn't notice a massive performance increase though). I recommend using the recently released SageAttention 2.0.1 if possible. Will also work for some other models like SD 1.5 which it didn't before.

**Comment by EnragedAntelope** [comment]
Thank you! FWIW Kijai nodes also do this so now two sage attn options for flux/other.
